ambry-frontend/src/main/java/com/github/ambry/frontend/PutBlobHandler.java

```diff
@@ -0,0 +1,405 @@
+        BlobInfo blobInfo = getBlobInfoFromRequest();
+        checkUploadRequirements(blobInfo.getBlobProperties());
+        securityService.processRequest(restRequest, securityProcessRequestCallback(blobInfo));
+      } catch (Exception e) {
+        finalCallback.onCompletion(null, e);
+      }
+    }
+
+    /**
+     * Parse {@link BlobInfo} from the request arguments. This method will also ensure that the correct account and
+     * container objects are attached to the request.
+     * @return the {@link BlobInfo} parsed from the request arguments.
+     * @throws RestServiceException if there is an error while parsing the {@link BlobInfo} arguments.
+     */
+    private BlobInfo getBlobInfoFromRequest() throws RestServiceException {
+      long propsBuildStartTime = System.currentTimeMillis();
+      accountAndContainerInjector.injectAccountAndContainerForPutRequest(restRequest,
+          frontendMetrics.putBlobMetricsGroup);
+      BlobProperties blobProperties = RestUtils.buildBlobProperties(restRequest.getArgs());
+      Container container = RestUtils.getContainerFromArgs(restRequest.getArgs());
+      if (blobProperties.getTimeToLiveInSeconds() + TimeUnit.MILLISECONDS.toSeconds(
+          blobProperties.getCreationTimeInMs()) > Integer.MAX_VALUE) {
+        LOGGER.debug("TTL set to very large value in POST request with BlobProperties {}", blobProperties);
+        frontendMetrics.ttlTooLargeError.inc();
+      } else if (container.isTtlRequired() && (blobProperties.getTimeToLiveInSeconds() == Utils.Infinite_Time
+          || blobProperties.getTimeToLiveInSeconds() > frontendConfig.maxAcceptableTtlSecsIfTtlRequired)) {
+        String descriptor = RestUtils.getAccountFromArgs(restRequest.getArgs()).getName() + ":" + container.getName();
+        if (frontendConfig.failIfTtlRequiredButNotProvided) {
+          throw new RestServiceException(
+              "TTL < " + frontendConfig.maxAcceptableTtlSecsIfTtlRequired + " is required for upload to " + descriptor,
+              RestServiceErrorCode.InvalidArgs);
+        } else {
+          LOGGER.debug("{} attempted an upload with ttl {} to {}", blobProperties.getServiceId(),
+              blobProperties.getTimeToLiveInSeconds(), descriptor);
+          frontendMetrics.ttlNotCompliantError.inc();
+          restResponseChannel.setHeader(RestUtils.Headers.NON_COMPLIANCE_WARNING,
+              "TTL < " + frontendConfig.maxAcceptableTtlSecsIfTtlRequired + " will be required for future uploads");
+        }
+      }
+      // inject encryption frontendMetrics if applicable
+      if (blobProperties.isEncrypted()) {
+        restRequest.getMetricsTracker()
+            .injectMetrics(frontendMetrics.postBlobMetricsGroup.getRestRequestMetrics(restRequest.isSslUsed(), true));
+      }
+      byte[] userMetadata = RestUtils.buildUserMetadata(restRequest.getArgs());
+      frontendMetrics.blobPropsBuildTimeInMs.update(System.currentTimeMillis() - propsBuildStartTime);
```

Comment:
Minor: this is likely copied code, but I doubt we need to time simple in-memory operations like this.

Reply:
I added another metric for the NamedBlob blobPropsBuildTime. This is what we have in PostBlobHandler. But if you think it's unnecessary, I can remove on both side?

Reply:
It's okay for now, we should probably go over the metrics at some point and remove the ones that are not useful.